File: Heaps/Last-Stone-Weight/last.py
import heapq
from typing import List

class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        pq = []
        n = len(stones)
        for i in range(n):
            heapq.heappush(pq,-stones[i])
        while len(pq)>1:
            firstStone = pq[0]
            heapq.heappop(pq)
            secondStone = pq[0]
            heapq.heappop(pq)
            diffStone = -abs(firstStone-secondStone)
            if diffStone:
                heapq.heappush(pq,diffStone)
            if len(pq)==0: return 0
        return -pq[0]

File: Heaps/Last-Stone-Weight/last.cpp
#include <vector>
#include <queue>
#include <cmath>
using namespace std;

class Solution {
public:
    int lastStoneWeight(vector<int>& stones) {
        priority_queue<int>pq;
        int n = stones.size();
        for(int i=0;i<n;i++){
            pq.push(stones[i]);
        }
        while(pq.size()>1){
            int firstStone = pq.top();
            pq.pop();
            int secondStone = pq.top();
            pq.pop();
            int diffStone = abs(firstStone-secondStone);
            if(diffStone){
                pq.push(diffStone);
            }
            if(pq.size()==0)return 0;
        }
        return pq.top();
    }
};